### PR TITLE
Add func to install nessessary packages for AWS

### PR DESCRIPTION
**Pull request by nathan-weinberg**
This PR adds a func to `cloud-instance.sh` that installs all necessary dnf-based packages for Fedora/RHEL

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: scripts/infra/cloud-instance.sh [hunk 1]
```diff
@@ -13,6 +13,17 @@ else
     exit 1
 fi
 
+ensure_system_pkgs() {
+    # shellcheck source=/etc/os-release
+    source /etc/os-release
```
Comment (danmcp) [inline scripts/infra/cloud-instance.sh:18]: This script is currently focused on operating on the cloud instance.  I am hesitant to also have it work on your local env.  Maybe this should be a sibling script in infra?

File: scripts/infra/cloud-instance.sh [hunk 2]
```diff
@@ -13,6 +13,17 @@ else
...
+    if [ $ID == "fedora" ] || [ $ID == "rhel" ]; then
+        PKG_LIST=(git gcc python-devel python-virtualenv krb5-devel openldap-devel aws)
+        sudo dnf install "${PKG_LIST[@]}"
```
Comment (danmcp) [inline scripts/infra/cloud-instance.sh:21]: Should this be with -y?  Or do you want confirmation?

**Review by danmcp — CHANGES_REQUESTED** [review]

File: scripts/infra/cloud-instance.sh (hunk 1 above)
Comment (nathan-weinberg) [reply scripts/infra/cloud-instance.sh:18]: I'm open to that for sure! I can separate this out in a followup commit

File: scripts/infra/cloud-instance.sh (hunk 2 above)
Comment (nathan-weinberg) [reply scripts/infra/cloud-instance.sh:21]: should have `-y` for sure

File: scripts/infra/README.md [hunk 3]
```diff
@@ -49,6 +50,8 @@ Either way, you'll need to edit the config to reference your cloud environment(s
 Example:
 
 ```bash
+# Ensure proper system packages are installed on Fedora/RHEL
+./cloud-instance.sh ec2 ensure-system-pkgs
```
Comment (danmcp) [inline scripts/infra/README.md:54]: setup-aws.sh

File: scripts/infra/setup-aws.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env -S bash -e
+# SPDX-License-Identifier: Apache-2.0
+
+ensure_system_pkgs() {
```
Comment (danmcp) [inline scripts/infra/setup-aws.sh:4]: Would this be better to create as a script that can have multiple sub commands.  This method for example should probably have rh in the name to leave room for other os types.

File: scripts/infra/setup-aws.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    # shellcheck source=/dev/null
+    source /etc/os-release
+    if [ "$ID" == "fedora" ] || [ "$ID" == "rhel" ]; then
```
Comment (danmcp) [inline scripts/infra/setup-aws.sh:7]: centos stream?

**Review by danmcp — CHANGES_REQUESTED** [review]

File: scripts/infra/setup-aws.sh (hunk 5 above)
Comment (nathan-weinberg) [reply scripts/infra/setup-aws.sh:7]: would you happen to know the `$ID` for CentOS? no worries if not, i can launch a VM and check

File: scripts/infra/setup-aws.sh (hunk 4 above)
Comment (nathan-weinberg) [reply scripts/infra/setup-aws.sh:4]: Yes I think that does make sense

File: scripts/infra/README.md (hunk 3 above)
Comment (nathan-weinberg) [reply scripts/infra/README.md:54]: good catch, thanks!

File: scripts/infra/setup-aws.sh (hunk 5 above)
Comment (tiran) [reply scripts/infra/setup-aws.sh:7]: Check for `[ "$ID" == "fedora" ] || [ "$ID_LIKE" == "fedora" ]`. Both CS and RHEL are like Fedora.

Comment (nathan-weinberg) [reply scripts/infra/setup-aws.sh:7]: thanks!

Comment (tiran) [reply scripts/infra/setup-aws.sh:7]: Correction, you need to check for `[[ "$ID_LIKE" =~ "fedora" ]]`

```
$ podman run --rm registry.access.redhat.com/ubi9/ubi:9.4 head -n6 /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="9.4 (Plow)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="9.4"
PLATFORM_ID="platform:el9"

$ podman run --rm quay.io/centos/centos:stream9 head -n6 /etc/os-release
NAME="CentOS Stream"
VERSION="9"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="9"
PLATFORM_ID="platform:el9"

```

**Comment by nathan-weinberg** [comment]
Tests from most recent changes:
```bash
[nathan@nathan-redhat infra (infra-work)]$ ./setup-aws.sh ensure-system-pkgs
Invalid system type: ensure-system-pkgs
Usage: ./setup-aws.sh <system-type> <command> [options]

System Types

    - rh (Red Hat-based system; Fedora, RHEL, CentOS)


Commands

    ensure-system-pkgs - Ensure your local system has the packages needed for AWS interactions


[nathan@nathan-redhat infra (infra-work)]$ ./setup-aws.sh ubuntu ensure-system-pkgs
Invalid system type: ubuntu
Usage: ./setup-aws.sh <system-type> <command> [options]

System Types

    - rh (Red Hat-based system; Fedora, RHEL, CentOS)


Commands

    ensure-system-pkgs - Ensure your local system has the packages needed for AWS interactions


[nathan@nathan-redhat infra (infra-work)]$ ./setup-aws.sh rh ensure-system-pkgs
Last metadata expiration check: 1:31:40 ago on Tue 02 Jul 2024 01:21:45 PM EDT.
Package git-2.45.0-1.fc38.x86_64 is already installed.
Package gcc-13.2.1-7.fc38.x86_64 is already installed.
Package python3-devel-3.11.9-2.fc38.x86_64 is already installed.
Package python3-virtualenv-20.21.1-7.fc38.noarch is already installed.
Package krb5-devel-1.21-3.fc38.x86_64 is already installed.
Package openldap-devel-2.6.6-1.fc38.x86_64 is already installed.
Package aws-2020-10.fc38.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```

File: scripts/infra/setup-aws.sh (hunk 5 above)
Comment (nathan-weinberg) [reply scripts/infra/setup-aws.sh]: fixed in most recent patchset

File: scripts/infra/setup-aws.sh [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        PKG_LIST=(git gcc python-devel python-virtualenv krb5-devel openldap-devel aws)
+        sudo dnf install "${PKG_LIST[@]}" -y
+    else
+        echo "neither system 'ID' nor 'ID_LIKE' is 'fedora' - did you mean to specify a 'rh' system?"
+    fi
+}
+
+rh() {
+    local cmdname=$1; shift
+    if [ "$(type -t "rh__$cmdname")" = "function" ] >/dev/null 2>&1; then
```
Comment (danmcp) [inline scripts/infra/setup-aws.sh:17]: Doesn't appear to be used

File: scripts/infra/setup-aws.sh [hunk 7]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env -S bash -e
```
Comment (danmcp) [inline scripts/infra/setup-aws.sh:1]: What if the script is named something like `local-setup`.  And the command was something like `ensure-aws-system-pkgs`.  So you could support other infras.

File: scripts/infra/setup-aws.sh (hunk 6 above)
Comment (danmcp) [inline scripts/infra/setup-aws.sh:17]: Doesn't seem to be used.  Although perhaps it should be since this method is rh specific.

**Review by danmcp — CHANGES_REQUESTED** [review]

File: scripts/infra/setup-aws.sh (hunk 7 above)
Comment (nathan-weinberg) [reply scripts/infra/setup-aws.sh:1]: I like it!

File: scripts/infra/setup-aws.sh (hunk 6 above)
Comment (nathan-weinberg) [reply scripts/infra/setup-aws.sh:17]: I figured this might be good to have already here for future additions

File: scripts/infra/local-setup.sh [hunk 8]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env -S bash -e
+# SPDX-License-Identifier: Apache-2.0
+
+ensure_aws_system_pkgs() {
```
Comment (danmcp) [inline scripts/infra/local-setup.sh:4]: This should be rh__ensure-aws-system-pkgs so it hits the first block in rh()

**Review by danmcp — CHANGES_REQUESTED** [review]

File: scripts/infra/local-setup.sh (hunk 8 above)
Comment (nathan-weinberg) [reply scripts/infra/local-setup.sh]: Done!

File: scripts/infra/README.md [hunk 9]
```diff
@@ -49,6 +50,8 @@ Either way, you'll need to edit the config to reference your cloud environment(s
 Example:
 
 ```bash
+# Ensure proper system packages are installed on Fedora/RHEL/CentOS
```
Comment (danmcp) [inline scripts/infra/README.md:53]: Probably need more of a warning that this is only applicable if your local env is Fedora/RHEL/CentOS

File: scripts/infra/README.md [hunk 10]
```diff
@@ -8,7 +8,8 @@ You'll need to setup the respective cloud CLI before using [cloud-instance.sh](c
 
 ### AWS
 
-- [aws](https://aws.amazon.com/cli/) (the AWS CLI)
+- You can ensure you have all the necessary local system packages with `setup-aws.sh`
```
Comment (danmcp) [inline scripts/infra/README.md:11]: Name changed to local-setup.sh

File: scripts/infra/README.md (hunk 9 above)
Comment (nathan-weinberg) [reply scripts/infra/README.md:53]: Added some addn text here, lmk what you think

File: scripts/infra/README.md [hunk 11]
```diff
@@ -49,6 +50,9 @@ Either way, you'll need to edit the config to reference your cloud environment(s
 Example:
 
 ```bash
+# Ensure proper system packages are installed on a Red Hat-based system (Fedora/RHEL/CentOS)
```
Comment (danmcp) [inline scripts/infra/README.md:53]: How about:

Ensure proper system packages are installed **if your local system is Red Hat-based**

File: scripts/infra/local-setup.sh [hunk 12]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env -S bash -e
+# SPDX-License-Identifier: Apache-2.0
+
+rh__ensure-aws-system-pkgs() {
+    # shellcheck source=/dev/null
+    source /etc/os-release
+    if [[ "$ID" == "fedora" ]] || [[ "$ID_LIKE" =~ "fedora" ]]; then
+        PKG_LIST=(git gcc python-devel python-virtualenv krb5-devel openldap-devel aws)
```
Comment (danmcp) [inline scripts/infra/local-setup.sh:8]: These packages:

gcc python-devel python-virtualenv krb5-devel openldap-devel

Wouldn't be needed by everyone.  It depends on how you are authenticating to aws.

Comment (nathan-weinberg) [reply scripts/infra/local-setup.sh:8]: Tbh I'm not sure where to go from here - this is really the meat of the PR - any suggestions?

Comment (danmcp) [reply scripts/infra/local-setup.sh:8]: It hit me too when I made the comment.  You could have a separate command for ensure aws sso auth pkgs.  But the comment would need to be pretty wordy to explain when/why you would need it.  We could easily have other use cases for this script to setup a local instructlab environment.  Maybe we need to settle on the second one before the commit will be worthwhile though.

**Review by leseb — CHANGES_REQUESTED** [review]
A few thoughts:

* Why is this new script needed? 
* Why isn't the logic part of `cloud-instance.sh`? I think I've read somewhere it used to be part of it but `cloud-instance.sh` is left untouched.
* The name of the script does not reflect what it is doing
* Please squash your commits

Thanks

**Comment by danmcp** [comment]
> * Why isn't the logic part of `cloud-instance.sh`? I think I've read somewhere it used to be part of it but `cloud-instance.sh` is left untouched.

I asked that it not be included in `cloud-instance.sh` because nothing else in cloud-instance.sh operates on your local environment.  I was worried about mixing the two and confusing the user about where changes were happening.

File: scripts/infra/local-setup.sh (hunk 12 above)
Comment (nathan-weinberg) [reply scripts/infra/local-setup.sh:8]: I had this thought - we install the base packages (just `git` and `aws) with a flag `-s` to install packages required for SSO/SAML authentication. I'll push up an implementation that we can discuss further on.

**Comment by nathan-weinberg** [comment]
@leseb In addition to @danmcp's comments, the crux of `Why is this new script needed?` is that right now a lot of the local setup required to use `cloud-instance.sh` is manual - I want to start automating these pieces.

File: scripts/infra/local-setup.sh [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        "${cmdname//-/_}" "rh"
+    else
+        echo "Invalid command: $cmdname"
+        show_usage
+        exit 1
+    fi
+}
+
+show_usage() {
+    echo "Usage: ./local-setup.sh <system-type> <command> [options]
```
Comment (leseb) [inline scripts/infra/local-setup.sh:45]:
```suggestion
    echo "Usage: "${BASH_SOURCE[0]}" <system-type> <command> [options]
```

**Review by leseb — CHANGES_REQUESTED** [review]

File: scripts/infra/local-setup.sh (hunk 13 above)
Comment (danmcp) [reply scripts/infra/local-setup.sh:45]: I made a similar change here: https://github.com/instructlab/instructlab/pull/1589/files

I don't think you want the "s

File: scripts/infra/README.md [hunk 14]
```diff
@@ -49,6 +50,9 @@ Either way, you'll need to edit the config to reference your cloud environment(s
 Example:
 
 ```bash
+# Ensure proper system packages are installed if your local system is Red Hat-based
+# If you are running a different OS you'll need to install these packages manually
+./local-setup.sh rh ensure-aws-system-pkgs
```
Comment (danmcp) [inline scripts/infra/README.md:55]: If this does install anything, you would have other steps to authenticate with aws still.  I am wondering whether it's best to just put a little more detail in Prereqs and not include the command here.

**Review by danmcp — CHANGES_REQUESTED** [review]

File: scripts/infra/local-setup.sh (hunk 13 above)
Comment (leseb) [reply scripts/infra/local-setup.sh:45]: I guess `$0` would work too, whatever we think is better rather than hardcoding the script name :)

Comment (nathan-weinberg) [reply scripts/infra/local-setup.sh:45]: Addressed in latest patchset

File: scripts/infra/README.md (hunk 14 above)
Comment (nathan-weinberg) [reply scripts/infra/README.md:55]: Addressed in latest patchset

**Review by leseb — APPROVED** [review]

**Review by danmcp — APPROVED** [review]

**Pull request merged**